### PR TITLE
Fixed error causing certain MadeWithLau recipes to crash

### PR DESCRIPTION
**Pull request by arpachuilo**
Hello,

Initially added here https://github.com/hhursev/recipe-scrapers/pull/572. Thanks for the quick response @jayaddison.

Some recipe instructions on their site don't have multiple "sub-steps" and therefore did not have a `div` tag which caused the scraper to crash for some recipes. So now I'm having the scraper fallback to just reading in that part as a whole.

I originally did this to match the sites original instructional format (as opposed to whole paragraphs) is there a preference on this when writing scrapers?

Also I just updated the original test for this site, should I have made it another test entirely?

Thanks!

### CODE REVIEW

File: recipe_scrapers/madewithlau.py [hunk 1]
```diff
@@ -56,7 +56,10 @@ def instructions(self):
         instructions = [
             p
             for pair in zip(headers, step_sets)
-            for p in [pair[0], *pair[1].find("div")]
+            for p in [
+                pair[0],
+                *(pair[1].find("div") if pair[1].find("div") else pair[1]),
```
Comment (jayaddison) [inline recipe_scrapers/madewithlau.py:61]: Possible simplification for this line (please check my logic on this):

```
*(pair[1].find("div") or pair[1]),
```

**Comment by jayaddison** [comment]
> Also I just updated the original test for this site, should I have made it another test entirely?

I think it's OK to update the existing test.  We do have some scrapers where multiple tests exist, and in a way that could be more robust - we don't have a policy on it, though.

(from experience I think it can _sometimes_ be a sign of software engineering project maturity if existing tests are never modified; if they are, it can be a sign that a major version bump or other breaking change is required - we're not that strict, though.  it'd also be tricky for us if multiple test-HTML files existed from multiple points during the history of a website.  maybe arguably that'd be good for us to encourage?  a longer discussion, perhaps :))

**Comment by jayaddison** [comment]
> I originally did this to match the sites original instructional format (as opposed to whole paragraphs) is there a preference on this when writing scrapers?

I'm not sure I understand this question, maybe I'm not familiar enough with the site - I see the paragraph-based format, I think - what's the original format you mention?

File: recipe_scrapers/madewithlau.py (hunk 1 above)
Comment (arpachuilo) [reply recipe_scrapers/madewithlau.py]: Yeah, that looks much cleaner and works the same. Thanks!

**Comment by arpachuilo** [comment]
> I think it's OK to update the existing test. We do have some scrapers where multiple tests exist, and in a way that could be more robust - we don't have a policy on it, though.

I noticed there were a few with multiple tests browsing the repo and I can see how that would be useful. I'll keep that in mind if an issue comes up again. Makes sense to me have multiple test.

> (from experience I think it can _sometimes_ be a sign of software engineering project maturity if existing tests are never modified; if they are, it can be a sign that a major version bump or other breaking change is required - we're not that strict, though. it'd also be tricky for us if multiple test-HTML files existed from multiple points during the history of a website. maybe arguably that'd be good for us to encourage? a longer discussion, perhaps :))

Yeah that would be a bit difficult to keep a historical test if a site did change. It makes sense to me the way it is structured now, it is nice and simple. I was thinking it could be useful to have tests that periodically check against the current version of the site for any major breaking changes beyond small edits, but if something did break someone might report it anyway.

**Comment by arpachuilo** [comment]
> I originally did this to match the sites original instructional format (as opposed to whole paragraphs) is there a preference on this when writing scrapers?

Yeah the difference is a bit subtle
Screen Shot 2022-07-11 at 14 08 25

The text on the left does not add any newlines between steps under the step header. Which is why I added the `pair[1].find("div")` originally which breaks them up into separate lines which fits the websites original format.

So I was just curious to which way would be preferred. Though looking at the text on the left again breaking them up with newlines is making more sense to me because there are some cases where the end of a sentence doesn't have any space between it and the next e.g. (`. . . end.Begin . . .` vs `. . . end. Begin . . . `

**Comment by jayaddison** [comment]
> Yeah that would be a bit difficult to keep a historical test if a site did change. It makes sense to me the way it is structured now, it is nice and simple. I was thinking it could be useful to have tests that periodically check against the current version of the site for any major breaking changes beyond small edits, but if something did break someone might report it anyway.

Yep, good thinking.  You might be interested to read #321 which covers similar ideas.

Also, we do have an 'online' scraper test mode (#245) which retrieves the URL from each test and runs the scraper against that; we don't use it within our continuous integration, though.

**Comment by jayaddison** [comment]
> The text on the left does not add any newlines between steps under the step header. Which is why I added the `pair[1].find("div")` originally which breaks them up into separate lines which fits the websites original format.
> So I was just curious to which way would be preferred.

Thanks for the explanation, that makes sense.

Here's an attempt at drafting some guidance (I think it would be useful - maybe we could document it within the repository):

`instructions` are ideally a series of newline-delimited actions that a chef would perform while preparing a recipe.  When writing a scraper, try to write your code so that -- as far as possible, based on the content contained on the webpage -- each line contains one "action" for a reader to follow.

(the rationale here: if someone were to build a text-to-voice app that would read out actions for people and allow them to "skip", "go forward", "go back", etc - it probably makes sense for each instruction to be effectively standalone)

(also, as a sidenote: the headings from this website would be a good fit for "grouped" instructions; see some previous discussion of this at https://github.com/hhursev/recipe-scrapers/issues/283#issuecomment-766411054)

**Comment by arpachuilo** [comment]
@jayaddison Thanks for all the information. Definitely interested in contributing more where I can.


> `instructions` are ideally a series of newline-delimited actions that a chef would perform while preparing a recipe. When writing a scraper, try to write your code so that -- as far as possible, based on the content contained on the webpage -- each line contains one "action" for a reader to follow.

That makes sense to me, think it's best to keep how it is now then. Is this ready to merge then?

> (part of the rationale here, by example: if someone were to build a text-to-voice app that would read out actions for people and allow them to "skip", "go forward", "go back", etc - it probably makes sense for each instruction to be effectively standalone)
> 
> (also, as a sidenote: the headings from this website would be a good fit for "grouped" instructions; see some previous discussion of this at [#283 (comment)](https://github.com/hhursev/recipe-scrapers/issues/283#issuecomment-766411054))

Great usability scenario and I'll definitely look more into grouping and read more of the codebase and discussion. Would be great addition for my use case as well.

**Pull request merged**